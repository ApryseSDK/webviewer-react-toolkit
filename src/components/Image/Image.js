'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
exports.Image = void 0;
var tslib_1 = require('tslib');
var classnames_1 = tslib_1.__importDefault(require('classnames'));
var react_1 = tslib_1.__importStar(require('react'));
var data_1 = require('../../data');
exports.Image = (0, react_1.forwardRef)(function(_a, ref) {
  var src = _a.src,
    pending = _a.pending,
    onRenderLoading = _a.onRenderLoading,
    onRenderFallback = _a.onRenderFallback,
    alt = _a.alt,
    className = _a.className,
    imgProps = tslib_1.__rest(_a, [
      'src',
      'pending',
      'onRenderLoading',
      'onRenderFallback',
      'alt',
      'className',
    ]);
  var sourceIsNotPromise = typeof src === 'string' || !src;
  var _b = (0, react_1.useState)(!sourceIsNotPromise),
    loading = _b[0],
    setLoading = _b[1];
  var _c = (0, react_1.useState)(sourceIsNotPromise ? src : undefined),
    source = _c[0],
    setSource = _c[1];
  var getSource = (0, react_1.useCallback)(function(srcGetter) {
    return tslib_1.__awaiter(void 0, void 0, void 0, function() {
      var fetchedSource, _a;
      return tslib_1.__generator(this, function(_b) {
        switch (_b.label) {
          case 0:
            setLoading(true);
            fetchedSource = undefined;
            _b.label = 1;
          case 1:
            _b.trys.push([1, 3, , 4]);
            return [
              4 /*yield*/,
              (0, data_1.futureableOrLazyToFuturable)(srcGetter),
            ];
          case 2:
            fetchedSource = _b.sent();
            return [3 /*break*/, 4];
          case 3:
            _a = _b.sent();
            return [3 /*break*/, 4];
          case 4:
            setLoading(false);
            setSource(fetchedSource || undefined);
            return [2 /*return*/];
        }
      });
    });
  }, []);
  (0, react_1.useEffect)(
    function() {
      if (sourceIsNotPromise) {
        setLoading(false);
        setSource(src || undefined);
        return;
      }
      getSource(src);
    },
    [getSource, sourceIsNotPromise, src],
  );
  var imageClass = (0, classnames_1.default)('ui__image', className);
  if (loading || pending)
    return react_1.default.createElement(
      react_1.default.Fragment,
      null,
      onRenderLoading === null || onRenderLoading === void 0
        ? void 0
        : onRenderLoading(),
    );
  if (!source)
    return react_1.default.createElement(
      react_1.default.Fragment,
      null,
      onRenderFallback === null || onRenderFallback === void 0
        ? void 0
        : onRenderFallback(),
    );
  return react_1.default.createElement(
    'img',
    tslib_1.__assign({}, imgProps, {
      alt: alt,
      src: source,
      className: imageClass,
      ref: ref,
    }),
  );
});
//# sourceMappingURL=Image.js.map
