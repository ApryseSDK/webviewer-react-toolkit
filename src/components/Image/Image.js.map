{"version":3,"file":"Image.js","sourceRoot":"","sources":["Image.tsx"],"names":[],"mappings":";;;;AAAA,kEAAoC;AACpC,qDAA0G;AAC1G,mCAA0E;AAwB7D,QAAA,KAAK,GAAG,IAAA,kBAAU,EAC7B,UAAC,EAAgF,EAAE,GAAG;IAAnF,IAAA,GAAG,SAAA,EAAE,OAAO,aAAA,EAAE,eAAe,qBAAA,EAAE,gBAAgB,sBAAA,EAAE,GAAG,SAAA,EAAE,SAAS,eAAA,EAAK,QAAQ,sBAA9E,6EAAgF,CAAF;IAC7E,IAAM,kBAAkB,GAAG,OAAO,GAAG,KAAK,QAAQ,IAAI,CAAC,GAAG,CAAC;IACrD,IAAA,KAAwB,IAAA,gBAAQ,EAAC,CAAC,kBAAkB,CAAC,EAApD,OAAO,QAAA,EAAE,UAAU,QAAiC,CAAC;IACtD,IAAA,KAAsB,IAAA,gBAAQ,EAClC,kBAAkB,CAAC,CAAC,CAAE,GAA0B,CAAC,CAAC,CAAC,SAAS,CAC7D,EAFM,MAAM,QAAA,EAAE,SAAS,QAEvB,CAAC;IAEF,IAAM,SAAS,GAAG,IAAA,mBAAW,EAAC,UAAO,SAA8C;;;;;oBACjF,UAAU,CAAC,IAAI,CAAC,CAAC;oBACb,aAAa,GAAG,SAAS,CAAC;;;;oBAEZ,qBAAM,IAAA,kCAA2B,EAAC,SAAS,CAAC,EAAA;;oBAA5D,aAAa,GAAG,SAA4C,CAAC;;;;;;oBAE/D,UAAU,CAAC,KAAK,CAAC,CAAC;oBAClB,SAAS,CAAC,aAAa,IAAI,SAAS,CAAC,CAAC;;;;SACvC,EAAE,EAAE,CAAC,CAAC;IAEP,IAAA,iBAAS,EAAC;QACR,IAAI,kBAAkB,EAAE;YACtB,UAAU,CAAC,KAAK,CAAC,CAAC;YAClB,SAAS,CAAE,GAA0B,IAAI,SAAS,CAAC,CAAC;YACpD,OAAO;SACR;QACD,SAAS,CAAC,GAAG,CAAC,CAAC;IACjB,CAAC,EAAE,CAAC,SAAS,EAAE,kBAAkB,EAAE,GAAG,CAAC,CAAC,CAAC;IAEzC,IAAM,UAAU,GAAG,IAAA,oBAAU,EAAC,WAAW,EAAE,SAAS,CAAC,CAAC;IAEtD,IAAI,OAAO,IAAI,OAAO;QAAE,OAAO,8DAAG,eAAe,aAAf,eAAe,uBAAf,eAAe,EAAI,CAAI,CAAC;IAC1D,IAAI,CAAC,MAAM;QAAE,OAAO,8DAAG,gBAAgB,aAAhB,gBAAgB,uBAAhB,gBAAgB,EAAI,CAAI,CAAC;IAChD,OAAO,0DAAS,QAAQ,IAAE,GAAG,EAAE,GAAG,EAAE,GAAG,EAAE,MAAM,EAAE,SAAS,EAAE,UAAU,EAAE,GAAG,EAAE,GAAG,IAAI,CAAC;AACvF,CAAC,CACF,CAAC"}