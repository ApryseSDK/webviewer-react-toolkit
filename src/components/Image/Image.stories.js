'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
exports.SrcPromiseReturnsFalsy = exports.SrcPromiseRejects = exports.WithSrcPromise = exports.Basic = void 0;
var tslib_1 = require('tslib');
var react_1 = tslib_1.__importDefault(require('react'));
var Image_1 = require('../Image');
var Spinner_1 = require('../Spinner');
var README_md_1 = tslib_1.__importDefault(require('./README.md'));
var addon_knobs_1 = require('@storybook/addon-knobs');
exports.default = {
  title: 'Components/Image',
  component: Image_1.Image,
  parameters: { readme: README_md_1.default },
};
var style = {
  height: 250,
  width: 250,
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
  backgroundColor: 'lightgray',
};
var IMAGE =
  'https://www.webfx.com/blog/images/cdn.designinstruct.com/files/582-how-to-image-placeholders/generic-image-placeholder.png';
var Basic = function() {
  return react_1.default.createElement(
    'div',
    { style: style },
    react_1.default.createElement(Image_1.Image, {
      src: IMAGE,
      onRenderLoading: function() {
        return react_1.default.createElement(Spinner_1.Spinner, null);
      },
      pending: (0, addon_knobs_1.boolean)('pending', false),
    }),
  );
};
exports.Basic = Basic;
var WithSrcPromise = function() {
  return react_1.default.createElement(
    'div',
    { style: style },
    react_1.default.createElement(Image_1.Image, {
      src: new Promise(function(res) {
        return setTimeout(function() {
          return res(IMAGE);
        }, 500);
      }),
      onRenderLoading: function() {
        return react_1.default.createElement(Spinner_1.Spinner, null);
      },
    }),
  );
};
exports.WithSrcPromise = WithSrcPromise;
var SrcPromiseRejects = function() {
  return react_1.default.createElement(
    'div',
    { style: style },
    react_1.default.createElement(Image_1.Image, {
      src: new Promise(function(rej) {
        return setTimeout(function() {
          return rej();
        }, 500);
      }),
      onRenderLoading: function() {
        return react_1.default.createElement(Spinner_1.Spinner, null);
      },
      onRenderFallback: function() {
        return 'Rejected source promise';
      },
      pending: (0, addon_knobs_1.boolean)('pending', false),
    }),
  );
};
exports.SrcPromiseRejects = SrcPromiseRejects;
var SrcPromiseReturnsFalsy = function() {
  return react_1.default.createElement(
    'div',
    { style: style },
    react_1.default.createElement(Image_1.Image, {
      src: new Promise(function(res) {
        return setTimeout(function() {
          return res('');
        }, 500);
      }),
      onRenderLoading: function() {
        return react_1.default.createElement(Spinner_1.Spinner, null);
      },
      onRenderFallback: function() {
        return 'Falsy source';
      },
      pending: (0, addon_knobs_1.boolean)('pending', false),
    }),
  );
};
exports.SrcPromiseReturnsFalsy = SrcPromiseReturnsFalsy;
//# sourceMappingURL=Image.stories.js.map
