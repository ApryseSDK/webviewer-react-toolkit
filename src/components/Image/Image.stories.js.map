{"version":3,"file":"Image.stories.js","sourceRoot":"","sources":["Image.stories.tsx"],"names":[],"mappings":";;;;AAAA,wDAA6C;AAC7C,kCAAiC;AACjC,sCAAqC;AACrC,kEAAiC;AACjC,sDAAiD;AAEjD,kBAAe,EAAE,KAAK,EAAE,kBAAkB,EAAE,SAAS,EAAE,aAAK,EAAE,UAAU,EAAE,EAAE,MAAM,qBAAA,EAAE,EAAE,CAAC;AAEvF,IAAM,KAAK,GAAkB;IAC3B,MAAM,EAAE,GAAG;IACX,KAAK,EAAE,GAAG;IACV,OAAO,EAAE,MAAM;IACf,cAAc,EAAE,QAAQ;IACxB,UAAU,EAAE,QAAQ;IACpB,eAAe,EAAE,WAAW;CAC7B,CAAC;AAEF,IAAM,KAAK,GACT,4HAA4H,CAAC;AAExH,IAAM,KAAK,GAAG,cAAM,OAAA,CACzB,uCAAK,KAAK,EAAE,KAAK;IACf,8BAAC,aAAK,IAAC,GAAG,EAAE,KAAK,EAAE,eAAe,EAAE,cAAM,OAAA,8BAAC,iBAAO,OAAG,EAAX,CAAW,EAAE,OAAO,EAAE,IAAA,qBAAO,EAAC,SAAS,EAAE,KAAK,CAAC,GAAI,CACzF,CACP,EAJ0B,CAI1B,CAAC;AAJW,QAAA,KAAK,SAIhB;AAEK,IAAM,cAAc,GAAG,cAAM,OAAA,CAClC,uCAAK,KAAK,EAAE,KAAK;IACf,8BAAC,aAAK,IAAC,GAAG,EAAE,IAAI,OAAO,CAAC,UAAC,GAAG,IAAK,OAAA,UAAU,CAAC,cAAM,OAAA,GAAG,CAAC,KAAK,CAAC,EAAV,CAAU,EAAE,GAAG,CAAC,EAAjC,CAAiC,CAAC,EAAE,eAAe,EAAE,cAAM,OAAA,8BAAC,iBAAO,OAAG,EAAX,CAAW,GAAI,CACvG,CACP,EAJmC,CAInC,CAAC;AAJW,QAAA,cAAc,kBAIzB;AAEK,IAAM,iBAAiB,GAAG,cAAM,OAAA,CACrC,uCAAK,KAAK,EAAE,KAAK;IACf,8BAAC,aAAK,IACJ,GAAG,EAAE,IAAI,OAAO,CAAC,UAAC,GAAG,IAAK,OAAA,UAAU,CAAC,cAAM,OAAA,GAAG,EAAE,EAAL,CAAK,EAAE,GAAG,CAAC,EAA5B,CAA4B,CAAC,EACvD,eAAe,EAAE,cAAM,OAAA,8BAAC,iBAAO,OAAG,EAAX,CAAW,EAClC,gBAAgB,EAAE,cAAM,OAAA,yBAAyB,EAAzB,CAAyB,EACjD,OAAO,EAAE,IAAA,qBAAO,EAAC,SAAS,EAAE,KAAK,CAAC,GAClC,CACE,CACP,EATsC,CAStC,CAAC;AATW,QAAA,iBAAiB,qBAS5B;AAEK,IAAM,sBAAsB,GAAG,cAAM,OAAA,CAC1C,uCAAK,KAAK,EAAE,KAAK;IACf,8BAAC,aAAK,IACJ,GAAG,EAAE,IAAI,OAAO,CAAC,UAAC,GAAG,IAAK,OAAA,UAAU,CAAC,cAAM,OAAA,GAAG,CAAC,EAAE,CAAC,EAAP,CAAO,EAAE,GAAG,CAAC,EAA9B,CAA8B,CAAC,EACzD,eAAe,EAAE,cAAM,OAAA,8BAAC,iBAAO,OAAG,EAAX,CAAW,EAClC,gBAAgB,EAAE,cAAM,OAAA,cAAc,EAAd,CAAc,EACtC,OAAO,EAAE,IAAA,qBAAO,EAAC,SAAS,EAAE,KAAK,CAAC,GAClC,CACE,CACP,EAT2C,CAS3C,CAAC;AATW,QAAA,sBAAsB,0BASjC"}