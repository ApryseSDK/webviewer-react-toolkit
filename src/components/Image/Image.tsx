import classnames from 'classnames';
import React, { forwardRef, ImgHTMLAttributes, useCallback, useEffect, useState, ReactNode } from 'react';
import { Omit } from '../../utils/typeUtils';

export interface ImageProps extends Omit<ImgHTMLAttributes<HTMLImageElement>, 'src'> {
  /**
   * The image source can be either a string, or a function that returns a
   * promise to get a string. If not given, will show as loading.
   */
  src?: string | (() => Promise<string>);
  /**
   * If given, will override the loading based on whether src is fetched.
   */
  loading?: boolean;
  /**
   * Specific classes for all of the internal elements. If you just wish to give
   * a class to image, you can use `className`.
   */
  classes?: { wrapper?: string; image?: string; overlay?: string };
  /**
   * Render out an element to be shown while src is loading.
   */
  onRenderLoading?: () => ReactNode;
  /**
   * Timeout to throttle getting the image. Only applies if `src` is a function.
   * @default 500;
   */
  throttleTimeout?: number;
}

export const Image = forwardRef<HTMLImageElement, ImageProps>(
  ({ src, loading, classes, onRenderLoading, throttleTimeout = 500, alt, className, ...imgProps }, ref) => {
    const [source, setSource] = useState<string | undefined>(typeof src !== 'function' ? src : undefined);

    const getSource = useCallback(async (srcGetter: () => Promise<string>) => {
      const fetchedSource = await srcGetter();
      setSource(fetchedSource);
    }, []);

    useEffect(() => {
      if (typeof src !== 'function') return setSource(src);
      const timeout = setTimeout(() => getSource(src), throttleTimeout);
      return () => clearTimeout(timeout);
    }, [src, getSource, throttleTimeout]);

    const wrapperClass = classnames('ui__base ui__image__wrapper', classes?.wrapper);

    const imageClass = classnames(
      'ui__image',
      { ['ui__image--loading']: loading ?? !source },
      classes?.image,
      className,
    );

    const overlayClass = classnames(
      'ui__image__overlay',
      {
        ['ui__image__overlay--loading']: loading ?? !source,
        ['ui__image__overlay--hide']: !onRenderLoading,
      },
      classes?.overlay,
    );

    return (
      <div className={wrapperClass}>
        <img {...imgProps} alt={alt} src={source} className={imageClass} ref={ref} />
        <div className={overlayClass}>{onRenderLoading?.()}</div>
      </div>
    );
  },
);
