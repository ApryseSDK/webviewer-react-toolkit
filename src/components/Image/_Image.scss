.ui__image__wrapper {
  max-height: 100%;
  position: relative;
  user-select: none;
  pointer-events: none;
}

.ui__image {
  max-width: 100%;
  max-height: 100%;
  display: block;
  border: 1px solid $blue-gray-1;
  transition: $focus-transition;
}

.ui__image__overlay {
  opacity: 0;
  pointer-events: none;

  > * {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    overflow: visible;
  }

  &--loading:not(.ui__image__overlay--hide) {
    opacity: 1;
    pointer-events: all;
  }
}
