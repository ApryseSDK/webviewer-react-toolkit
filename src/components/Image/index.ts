export { Image as default, ImageProps } from './Image';
