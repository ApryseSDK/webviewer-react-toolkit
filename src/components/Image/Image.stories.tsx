import { text } from '@storybook/addon-knobs';
import React from 'react';
import Image from '../Image';
import docs from './README.md';

export default { title: 'Image', parameters: { info: docs } };

export const basic = () => <Image src={text('someProp', 'Hello, World!')} />;
