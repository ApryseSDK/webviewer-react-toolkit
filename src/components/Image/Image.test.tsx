import { shallow } from 'enzyme';
import React from 'react';
import Image from '../Image';

describe('Image component', () => {
  it('renders its contents', () => {
    const image = shallow(<Image />);
    expect(image.find('.ui__image').length).toEqual(1);
  });

  it('snapshot renders default image', () => {
    const image = shallow(<Image />);
    expect(image).toMatchSnapshot();
  });
});
