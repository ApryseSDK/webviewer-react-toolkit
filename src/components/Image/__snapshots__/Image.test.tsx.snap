// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image component snapshot renders default image 1`] = `
<div
  className="ui__base ui__image__wrapper"
>
  <img
    className="ui__image ui__image--loading"
  />
  <div
    className="ui__image__overlay ui__image__overlay--loading ui__image__overlay--hide"
  />
</div>
`;
